src/lib/device-logs/backends/loki.ts

```diff
@@ -0,0 +1,172 @@
+	}
+
+	public get available(): boolean {
+		// TODO: check /ready endpoint?
+		return true;
+	}
+
+	public async history(ctx: LogContext, count: number): Promise<DeviceLog[]> {
+		const [response, body] = await requestAsync({
+			url: `http://${LOKI_HOST}:${LOKI_PORT}/loki/api/v1/query_range`,
+			headers: {
+				'X-Scope-OrgID': ctx.belongs_to__application,
+			},
+			method: 'GET',
+			qs: {
+				query: this.getDeviceQuery(ctx),
+				limit: Number.isFinite(count) ? count : 1000,
+			},
+			json: true,
+		});
+		if (response.statusCode === 200) {
+			return this.fromLokiStreamsToDeviceLogs(body.data.result);
+		} else {
+			throw new BadRequestError('Query failed');
+		}
+	}
+
+	public async publish(ctx: LogWriteContext, logs: DeviceLog[]): Promise<any> {
+		return requestAsync({
+			url: `http://${LOKI_HOST}:${LOKI_PORT}/loki/api/v1/push`,
+			headers: {
+				'X-Scope-OrgID': ctx.belongs_to__application,
+			},
+			method: 'POST',
+			json: {
+				streams: logs.map((log) => this.toLokiStream(ctx, log)),
+			},
+		});
+	}
+
+	public subscribe(ctx: LogContext, subscription: Subscription) {
+		const key = this.getKey(ctx);
+		if (!this.webSockets.has(key)) {
+			const tailUri =
+				`ws://${LOKI_HOST}:${LOKI_PORT}/loki/api/v1/tail` +
+				`?query=${encodeURIComponent(this.getDeviceQuery(ctx))}&start=${String(
+					Date.now(),
+				)}000000`;
+			const ws = new WebSocket(tailUri, {
+				headers: { 'X-Scope-OrgID': ctx.belongs_to__application },
+			});
+			ws.on('message', (data: string) => {
+				try {
+					const messageObj = JSON.parse(data);
+					if (messageObj.streams) {
+						this.fromLokiStreamsToDeviceLogs(messageObj.streams).map(
+							(log: DeviceLog) => {
+								this.subscriptions.emit(key, log);
+							},
+						);
+					}
+				} catch (err) {
+					captureException(err, 'Unexpected message received from Loki tail');
+				}
+			});
+			ws.on('close', () => {
+				this.webSockets.delete(key);
+			});
+			this.webSockets.set(key, ws);
+		}
+		this.subscriptions.on(key, subscription);
+	}
+
+	public unsubscribe(ctx: LogContext, subscription: Subscription) {
+		const key = this.getKey(ctx);
+		this.subscriptions.removeListener(key, subscription);
+		if (!this.webSockets.has(key)) {
+			const ws = this.webSockets.get(key);
+			if (ws) ws.close();
+		}
+	}
+
+	private getDeviceQuery(ctx: LogContext) {
+		return `{device_id="${ctx.id}"}`;
+	}
+
+	private getKey(ctx: LogContext, suffix = 'logs') {
+		return `device:${ctx.id}:${suffix}`;
+	}
+
+	private fromLokiStreamsToDeviceLogs(streams: LokiStream[]): DeviceLog[] {
+		return streams.flatMap(this.fromLokiStream);
+	}
+
+	private fromLokiStream(stream: LokiStream): DeviceLog[] {
+		try {
+			return stream.values.map(([, message]) => {
+				const log = JSON.parse(message);
+				if (log.version !== VERSION) {
+					throw new Error(
+						`Invalid Redis serialization version: ${JSON.stringify(log)}`,
+					);
+				}
+				if (log.serviceId === null) {
+					delete log.serviceId;
+				}
```

Comment:
When will `log.serviceId` be null? It'd make more sense to me to do this deletion when receiving the log instead

Reply:
I have removed this. For Redis backend `avro` would apply a default which was null but for Loki to `serviceId` will not be part of the JSON string if undefined so deleting it is not necessary.